crates/ruff_db/src/diagnostic/mod.rs

```diff
@@ -0,0 +1,528 @@
+/// an invalid type, then the span should point to the specific argument and
+/// not to the entire function call.
+///
+/// Messages attached to annotations should also be as brief and specific as
+/// possible. Long messages could negative impact the quality of rendering.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Annotation {
+    /// The span of this annotation, corresponding to some subsequence of the
+    /// user's input that we want to highlight.
+    span: Span,
+    /// An optional message associated with this annotation's span.
+    ///
+    /// When present, rendering will include this message in the output and
+    /// draw a line between the highlighted span and the message.
+    message: Option<Box<str>>,
+    /// Whether this annotation is "primary" or not. When it isn't primary, an
+    /// annotation is said to be "secondary."
+    is_primary: bool,
+}
+
+impl Annotation {
+    /// Create a "primary" annotation.
+    ///
+    /// A primary annotation is meant to highlight the "locus" of a diagnostic.
+    /// That is, it should point to something in the end user's input that is
+    /// the subject or "point" of a diagnostic.
+    ///
+    /// A diagnostic may have many primary annotations. A diagnostic may not
+    /// have any annotations, but if it does, at least one _ought_ to be
+    /// primary.
+    pub fn primary(span: Span) -> Annotation {
+        Annotation {
+            span,
+            message: None,
+            is_primary: true,
+        }
+    }
+
+    /// Create a "secondary" annotation.
+    ///
+    /// A secondary annotation is meant to highlight relevant context for a
+    /// diagnostic, but not to point to the "locus" of the diagnostic.
+    ///
+    /// A diagnostic with only secondary annotations is usually not sensible,
+    /// but it is allowed and will produce a reasonable rendering.
+    pub fn secondary(span: Span) -> Annotation {
+        Annotation {
+            span,
+            message: None,
+            is_primary: false,
+        }
+    }
+
+    /// Attach a message to this annotation.
+    ///
+    /// An annotation without a message will still have a presence in
+    /// rendering. In particular, it will highlight the span association with
+    /// this annotation in some way.
+    ///
+    /// When a message is attached to an annotation, then it will be associated
+    /// with the highlighted span in some way during rendering.
+    pub fn message<'a>(self, message: impl std::fmt::Display + 'a) -> Annotation {
+        let message = Some(message.to_string().into_boxed_str());
+        Annotation { message, ..self }
+    }
+}
+
+/// A string identifier for a lint rule.
+///
+/// This string is used in command line and configuration interfaces. The name should always
+/// be in kebab case, e.g. `no-foo` (all lower case).
+///
+/// Rules use kebab case, e.g. `no-foo`.
+#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct LintName(&'static str);
```

Comment:
It's somewhat unfortunate that git thinks everything below here is new when it isn't. It would be nice to keep the history